src/createRecords.ts

```diff
@@ -0,0 +1,64 @@
+import AirtableSDK from "airtable"
+
+import { SchemaZodT } from "./types/schema"
+import { SelectQueryParamsT } from "./types/queryParams"
+
+export interface CreateRecords<T extends SchemaZodT> {
+	tableId: string
+	baseId: string
+	schema: T
+	query?: SelectQueryParamsT
+	records: Array<T>
+}
+
+export interface CreateRecordsWithAirtable<T extends SchemaZodT>
+	extends CreateRecords<T> {
+	airtable: AirtableSDK
+}
+export interface CreateRecordsWithApiKey<T extends SchemaZodT>
+	extends CreateRecords<T> {
+	apiKey: string
+}
+
+//TODO should we change this to an upsert operation?
+export function createRecords<T extends SchemaZodT>({
+	tableId,
+	schema,
+	baseId,
+	records,
+	...otherArgs
+}: CreateRecordsWithAirtable<T> | CreateRecordsWithApiKey<T>) {
+	let airtable: AirtableSDK
+	if ("airtable" in otherArgs) {
+		airtable = otherArgs.airtable
+	} else {
+		airtable = new AirtableSDK({ apiKey: otherArgs.apiKey })
+	}
+
+	const parseddata = records.map((r) => schema.safeParse(r))
+
+	const data = parseddata
+		.flatMap((r) => (r.success === true ? r : []))
+		.map((r) => r.data)
+
+	const failedValidation = parseddata.flatMap((r) =>
+		r.success === false ? r : []
+	)
+
+	const batchSize = 10
+	const batches = Array.from(
+		{ length: Math.ceil(data.length / batchSize) },
+		(_, i) => data.slice(i * batchSize, i * batchSize + batchSize)
+	)
+
+	return Promise.all(
+		batches.map((batch) => {
+			return airtable.base(baseId).table(tableId).create(batch)
+		})
+	).then((results) => {
+		return {
+			data: results,
+			failedValidation,
+		}
+	})
```

Comment:
@iainkirkpatrick @hikurangi here is an example of the behaviour I want the function to have, but I am unsure of the return type.

i.e. the code has successfully executed with batching on x number of records and returns them including the new airtableRecordIds which may need to be persisted somewhere. It also returns the records that failed validation. Its up to the Developer then to work out what to do with these.

Reply:
IMO the result I'd want is the same array of records mapped, where each is now either the result of a create or an error - that means the caller can iterate over the array and decide what to do with them. Preserving the original order / index of records can sometimes be quite useful.